### DISCUSSION TITLE
Limit the amount of exceptions per function

### DISCUSSION THREAD

**Discussion by veewee** [azjezz/psl#182] [Ideas]
Take following example:

```php
namespace Psl\Regex;

/**
 * @throws Exception\InvalidPatternException If $pattern is invalid.
 * @throws Exception\RuntimeException In case of an unexpected error.
 * @throws Psl\Exception\InvariantViolationException If $limit is negative, or equal to 0.
 */
function split(string $subject, string $pattern, ?int $limit = null): array
```

This function can throw 3 different types of exceptions:

- `InvalidPattern`: It can indeed throw that, but that should (preferably) not happen outside of development.
- `RuntimeExcetion` : shit happens ...
- `InvariantViolantionException` : This can happen based on the limit attribute.

Now if I want to use the split function inside my code, I need to add a supress for both invalid pattern and invariant violation, since it does not apply to me.

Most of the time, I am only interested in dealing with the `RuntimeException`.
The other errors are mostly things that occur during development, but don't really need to buble upstream.

The 3 exceptions don't really have a main interface outside the most generic `Psl\Exception\ExceptionInterface`.
I don't really have an answer to this problem, so is there something we could do about this from your perspective?

What would be nice - but maybe out of scope for this project is:

- regex pattern validation in psalm instead of during runtime (not ideal if you are not using psalm in your project though)
- conditional throw statements in psalm. Like : `@throws-if $limit <= 0 InvariantViolantionException`

**Comment by azjezz** [comment]
Hm, i doubt `@throws-if` will make it into psalm, but, maybe we can add an entry to psalm.xml to list exceptions to ignore, because most of the time, you want to ignore InvariantViolation, InvariantViolation only happens when the end users violates the API ( passes a negative limit for example, when the the function states that it shouldn't be negative ).

as for the other exceptions, PSL functions should only throw exceptions from within it's component namespace, and each component should contain it's own ExceptionInterface that extends the base ExceptionInterface, so you can catch `Psl\Regex\ExceptionInterface` in this case instead of `Psl\Regex\RuntimeException` ( if you are sure pattern is always valid ).

**Reply by veewee** [reply]
Ah ok, did not know it was possible to ignore these errors:

https://psalm.dev/docs/running_psalm/configuration/#ignoreexceptions

I think that covers the issues I am dealing with at the moment.
Gonna add these to my project for now:

```xml
  <ignoreExceptions>
      <class name="Psl\Regex\Exception\InvalidPatternException" />
      <class name="Psl\Exception\InvariantViolationException" />
  </ignoreExceptions>
```
